aztec/src/main/kotlin/org/wordpress/aztec/watchers/EnterPressedWatcher.kt

```diff
@@ -0,0 +1,49 @@
+
+package org.wordpress.aztec.watchers
+
+import android.text.Editable
+import android.text.SpannableStringBuilder
+import android.text.TextWatcher
+import org.wordpress.aztec.AztecText
+import org.wordpress.aztec.Constants
+import java.lang.ref.WeakReference
+
+class EnterPressedWatcher(aztecText: AztecText) : TextWatcher {
+
+    private val aztecTextRef: WeakReference<AztecText?> = WeakReference(aztecText)
+    private var textBefore : SpannableStringBuilder? = null
+    private var start: Int = -1
+
+    override fun beforeTextChanged(text: CharSequence, start: Int, count: Int, after: Int) {
+        val aztecText = aztecTextRef.get()
+        if (aztecText?.getAztecKeyListener() != null && !aztecText.isTextChangedListenerDisabled()) {
+            // we need to make a copy to preserve the contents as they were before the change
```

Comment:
Given we really only need to extend the behavior of detecting `Enter` for our Gutenberg purposes for now, I wonder if there is a way for us to know we're in "Gutenberg mode" at this level, and just only do this copy / enable (or even install) the `EnterPressedWatcher` when needed. The reason behind this is I'm thinking we may have long content on usual Aztec, and making a copy in memory will probably become a problem (as opposed to Gutenberg blocks which should be shorter).
I understand that if the listener is not set, then this won't be run, but it also means the `onBackspaceKey` will have to be disabled. If these two are closely related to Gutenberg, should we then make that explicit? And if on the contrary these 2 are not closely related, can we make a separation by splitting the interface into 2 different ones (one for `onBackspaceKey` and another one for `onEnterKey`)?
What do you think?

Reply:
Let me clarify my thoughts: I think we should only [`install()` the `EnterPressedWatcher`](https://github.com/wordpress-mobile/AztecEditor-Android/pull/763/files#diff-7ca71cd1f009ab8438e0901205001b14R507) when the `onAztecKeyListener` is set (so the watcher is not run every time). Also, we should probably split the interface into 2 different interfaces.

Reply:
Yes, we can `install` the enter watcher when the key listener is set on Aztec, but it will be added at the end of the list of watchers, instead we want it at the beginning to prevent other watchers to modify the text before it.
Most probably there is no difference at the moment, but who knows in the future?
Also, this enter.key watcher is always installed, but nothing is copied in memory if the key-listener is null. It's only copied when the listener is set in the AztecWrapper for React Native. 
Same is for the backspace listener, it does intercept the key pressed, only when the listener is set.

Note that Aztec has already a big pile of watchers, that do copy the current content in memory each time the user presses a key.


> Also, we should probably split the interface into 2 different interfaces.

We had 2 distinct interfaces, but merged it in one, since it's all about intercepting a key pressed event. There is no big difference for me, but i prefer to have 1 interface and 2 or 3 methods, each for a different key.

Reply:
> but it will be added at the end of the list of watchers, instead we want it at the beginning to prevent other watchers to modify the text before it.

I see what you mean. Can we then add a comment in code, to make sure to warn any other developer seeing that does not change the place of the `install()` call and leaves it as the very first TextWatcher to be added?

> nothing is copied in memory if the key-listener is null. 

Gotcha,  thanks for the clarification @daniloercoli.

> Note that Aztec has already a big pile of watchers, that do copy the current content in memory each time the user presses a key.

A clarification here: while it's true there are a lot of TextWatchers (and some recursion in it), AFAICT the only ones that copy content are the HistoryLogger (which seems to be the heaviest in terms of memory usage) and the ObservationQueue but this latter one only copies at the event (textChanged) level and for a very limited size (then longest is as long as a known pattern size is, which is max 4 events currently). The other watchers only perform specific in-line modifications on spans as long as I can see.

Also, maybe we should add a comment (or lint warning?) to the interface so the developer knows that by setting a listener for such interface a copy of the text will be generated on each `enter` pressed

Reply:
> Can we then add a comment in code, to make sure to warn any other developer seeing that does not change the place of the install() call and leaves it as the very first TextWatcher to be added?

> Also, maybe we should add a comment (or lint warning?) to the interface so the developer knows that by setting a listener for such interface a copy of the text will be generated on each enter pressed

Right, will add a comment on both places. 👍 

> A clarification here: while it's true there are a lot of TextWatchers (and some recursion in it), AFAICT the only ones that copy content are the HistoryLogger (which seems to be the heaviest in terms of memory usage) and the ObservationQueue but this latter one only copies at the event (textChanged) level and for a very limited size (then longest is as long as a known pattern size is, which is max 4 events currently). The other watchers only perform specific in-line modifications on spans as long as I can see.

I quickly checked yesterday, when started working on this PR, and it seems there are other watchers that does copy the content. `EndOfParagraphMarkerAdder` is one of those for example.

Reply:
> I quickly checked yesterday, when started working on this PR, and it seems there are other watchers that does copy the content. EndOfParagraphMarkerAdder is one of those for example.

Oh thanks for pointing that out. Just saw that 👍

Reply:
`SuggestionWatcher` and `InlineTextWatcher` are other examples of watchers that hold a  copy of the whole text in `beforeTextChanged`.

Reply:
yep I see `TextChangedEvent` holding the copy there 👍 . I imagine we could have 1 instance of the TextChangedEvent and pass it through to all listeners 🤔 (Not in this PR, just thinking aloud)

Reply:
Yes, we can probably open an issue for this to keep track of it.